lib/prometheus/client/data_stores/direct_file_store.rb

```diff
@@ -196,6 +196,8 @@ def initialize(filename, readonly = false)
             open_file(filename, readonly)
             @used = @f.read(4).unpack('l')[0] if @capacity > 0
 
+            return if readonly
+
             if @used > 0
```

Comment:
I like this idea. *However*, it make usage of this object dangerous. 
Basically, this means that if you init with readonly, the *only* method you can call is `all_values`
If you call anything else, you'll corrupt the file.

This happens to be fine right now, because in the only case where we use readonly, that's the only method we calll. But i'm worried of not being aware of this if we (or anyone else) ever modify code down the line, this makes it quite brittle / dangerous...

Can we have some sort of `@file_map_loaded?` instance var that gets set by `read_all_values`? Then, all the other methods would have to check it and call `read_all_values` if that var is false.

It's quite dirty, but it'd help with the memory allocation.
Alternatively, we can change `all_values` so that it *doesn't* call `read_all_values`, and uses `@positions` instead? It'll make its code more complex but accomplish what we're trying to do here?

Reply:
I agree. Was focused on the code path followed by exporting/formatting and didn't realise that other methods would break if `@positions` is not indexed at creation time.

On that point, usage the `readonly` flag might be a bit confusing here in that it still allows write operations to go through, but that's outside the topic of this PR.

So, I pushed a second commit that reverts my previous change (`@positions` is again indexed at creation time) and uses `@positions` in `all_values` to seek across all values. Things look good, but the peak memory reduction I got in the benchmark attached to this PR is not as much exciting as it was before:

```
# gc_production_branch_do_not_push
# Total allocated: 127.99 MB (2231891 objects)
# Total retained:  139.68 kB (609 objects)

# Second commit in this PR
Total allocated: 95.93 MB (1566004 objects)
Total retained:  139.51 kB (608 objects)
```

This is because, while not reading file content twice, we're still creating more short-living objects.

However, this second change leaves `read_all_values` used only at creation time, so it opens the doors to another optimisation: I pushed a 3rd commit that rewrites `read_all_values` in order to yield key-position pairs and avoid reading metric values at creation time.

This seems to bring numbers back to the original figures:

```
# gc_production_branch_do_not_push
# Total allocated: 127.99 MB (2231891 objects)
# Total retained:  139.68 kB (609 objects)

# Second commit in this PR
Total allocated: 95.93 MB (1566004 objects)
Total retained:  139.51 kB (608 objects)

# Third commit in this PR
Total allocated: 86.27 MB (1334209 objects)
Total retained:  140.18 kB (612 objects)
```

Reply:
I like this new version! 
I have one further proposal. `read_all_values` is now only called from one place... Instead  of accepting a block and yielding to it for each value... You could just populate `@positions` from inside `read_all_values` and remove that indirection.

Reply:
@dmagliola pushed another commit with the changes you suggested. Would you mind having another pass?

CC @Sinjo